### ISSUE TITLE
Allow a default value for TXP variables

### ISSUE THREAD

**Issue by cara-tm** [textpattern/textpattern#1646]
### Is your feature request related to a problem?

No. A kind of enhancement.

### What is the feature?

All `custom_field` fields allow a default value (which is very useful!). What do you think about adding such an alternative for variables too?

**Comment by Bloke** [comment]
It already does. `default` is a global attribute :)

But, I think there might be a bug. Steps:

1) Create a new Form with this content:
``` html
<p>
Here is the article excerpt: <txp:excerpt escape="trim" />

<p>Here is the assignment of it (could also use a container with the same result)...

<txp:variable name='checkcerpt' default="No excerpt here" value='<txp:excerpt escape="trim" />' />

<p>Stored value: <txp:variable name='checkcerpt' />
```

2) Create an article with no excerpt set. Assign the above form as the override form for ease of testing.
3) View the article

Expected:
Here is the article excerpt:
Here is the assignment of it (could also use a container with the same result)...
Stored value: No excerpt here.

Actual:
Here is the article excerpt:
Here is the assignment of it (could also use a container with the same result)... No excerpt here
Stored value: No excerpt here.

Notice the default value is echoed when the assignment is made. Adding `output="0"` to the variable tag makes no difference.

If you assign a value to your article's Excerpt field and view the article again, the article's excerpt is displayed (correctly) above and below the assignment, but the assignment default value is still output to the screen.

This doesn't feel like it should be by design. Just me? Paging @bloatware :)

**Comment by cara-tm** [comment]
Yep Stef. I tried too.
Unfortunately, the variable is doubled and displayed by default...

**Comment by bloatware** [comment]
It's kinda by design. Global `default` attribute applies if the tag outputs nothing, which is the case of
```
<txp:variable name='checkcerpt' value='<txp:excerpt escape="trim" />' />
```
To make it output what you expect, modify the form like
```
<p>Here is the article excerpt: <txp:excerpt escape="trim" /></p>
<p>Here is the assignment of it (could also use a container with the same result)...</p>
<txp:variable name='checkcerpt' value='<txp:excerpt escape="trim" default="No excerpt here" />' />
<p>Stored value: <txp:variable name='checkcerpt' /></p>
```

**Comment by Bloke** [comment]
Okay, that's cool in this case. That makes sense! Thanks for the clarification, Oleg.

**Comment by cara-tm** [comment]
I'm sorry, but this isn't a direct assignment of a default value to a variable, but rather the one from the `excerpt`:

`<txp:variable name='checkcerpt' value='<txp:excerpt escape="trim" default="No excerpt here" />' />`

I'm not the only one who uses a form to store `config` variables for an entire website. What if one of them is empty?

**Comment by Bloke** [comment]
That is a good point. But I can also see Oleg's view that the `<txp:variable />` tag is an _output_ tag if used without a value. In that sense, you can assign it somewhere near the top of your page and then - at the point you want to output it - specify what should be displayed _if the content of the variable is empty_. It's not what should be _stored_ in the variable.

Consider:

``` html
<!-- Assignment -->
<txp:variable name="primary_section" value="about"  />
```

Output: _empty_

``` html
<!-- Display: value attribute is ignored -->
<txp:variable name="primary_section" value="about" default="articles" />
```

Output: `articles`

This demonstrates it even more clearly:

``` html
<!-- variable has a value so is output by the 2nd tag -->
<txp:variable name="primary_section" value="about" />
<txp:variable name="primary_section" default="articles" />
```
Output: `about`

``` html
<!-- variable has no value so default value is output by the 2nd tag -->
<txp:variable name="primary_section" value="" />
<txp:variable name="primary_section" default="articles" />
```
Output: `articles`

tl;dr: if the content of the variable is empty for whatever reason during the assignment stage, you can check that and output a default value _when you come to use that variable_.

I admit it is a little confusing so we should definitely add clear examples to the tag docs.

**Comment by cara-tm** [comment]
Ok. That makes sense. Thank you Stef to clarify. Please, close this request.

**Comment by bloatware** [comment]
I agree that in its current form this is useless and certainly unused:
```
<txp:variable name="primary_section" value="about" default="articles" />
```
So we could alter `default` in `<txp:variable />` to make it more clever.

**Comment by Bloke** [comment]
On its own, that example is quite rubbish. I mean, what would anyone expect it to do? Default would clearly be ignored. But combined with _dynamic_ (tag in tag or container) content for the `value` then a `default` is useful to be able to _assign_ the default value to the variable if the content is empty.

There are edge cases. If you're setting the value as a container, then it will never be 'empty' unless you butt the content up to the tag pair, or use `trim`, so that makes this shortcut a little less intuitive. As `default` is global, I presume this will affect all uses of it, across all tags, in a similar way?

The current "default on output only" restriction is fine, imo, if potentially head-scratchy as cara-tm found. If it can be made more intelligent, then great. I've [documented the example above](https://docs.textpattern.com/tags/variable) for now so it's much clearer how `default` interacts with the `value`. We'll of course need to revisit that if we change its behaviour somehow.

**Comment by bloatware** [comment]
Global `default` checks the trimmed content anyway, so we would adopt the same behaviour here. Actually, `escape` in `<txp:variable />` already acts in a similar way, depending on tag's 'setting' or 'getting' form.

**Comment by cara-tm** [comment]
Or we can achieve a simple test on variable like this (the result change by removing "Bob"):

    <txp:variable name="my_test" value="Bob" />

    <p>Output: <txp:variable name="my_test" value='<txp:evaluate test=''<txp:variable name="my_test" />''>New Value<txp:else/><txp:variable name="my_test" /></txp:evaluate>' /> <txp:variable name="my_test" /></p>

**Comment by bloatware** [comment]
You should be able to shorten the second line as
```
<txp:variable name="my_test" value='<txp:variable name="my_test" default="New value" />' />
```
Or, if you need to assign a default value only if the variable is unset:
```
<txp:variable name="my_test" value>New value</txp:variable>
```
But, since
```
<txp:variable name="my_test" value='<txp:maybe_empty />' default="New value" />
```
construction is currently useless anyway, we could give it a sense.

**Comment by bloatware** [comment]
[Now](https://github.com/textpattern/textpattern/commit/f12512b1796159be6933450304635771169ffae2) `default` value is *assigned* in 'set' mode if trimmed `value` (or content) is empty:
```
<!-- the variable is assigned "New value" if <txp:maybe_empty /> is empty -->
<txp:variable name="my_test" value='<txp:maybe_empty />' default="New value" />

<!-- the variable is assigned "New value" if its current value is empty -->
<txp:variable name="my_test" value default="New value" />
```
Add `output` attribute if you want to display a variable after setting it.

**Comment by cara-tm** [comment]
Parse error: syntax error, unexpected '{' in C:\MAMP\htdocs\website\textpattern\publish\taghandlers.php on line 5135

**Comment by bloatware** [comment]
Bizarre, there is no `{` on [line 5135](https://github.com/textpattern/textpattern/blob/f12512b1796159be6933450304635771169ffae2/textpattern/publish/taghandlers.php#L5135).

**Comment by cara-tm** [comment]
Yes. That's very strange. Did you test this stage of the distribution?

**Comment by bloatware** [comment]
I've tested my local copy, and [4.8.5 demo](https://dev-demo.textpattern.co/4.8.5/) seems to work ok too. What is the code you test?

**Comment by cara-tm** [comment]
Ok. So, it's fine (maybe I made a mistake).

**Comment by cara-tm** [comment]
Fine. Thanks. It's perfect! Can be close, now.

**Comment by bloatware** [comment]
TODO: rewrite the corresponding docs example when 4.8.5 is released.
